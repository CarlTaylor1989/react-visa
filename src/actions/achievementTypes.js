export const SET_COMPLETED_ACHIEVEMENT = 'SET_COMPLETED_ACHIEVEMENT';
export const RESTORE_ACHIEVEMENTS = 'RESTORE_ACHIEVEMENTS';

export const UPDATE_PRODUCT_ACHIEVEMENTS = 'UPDATE_PRODUCT_ACHIEVEMENTS';
export const UPDATE_PROPOSAL_ACHIEVEMENTS = 'UPDATE_PROPOSAL_ACHIEVEMENTS';
export const UPDATE_SCORED_ACHIEVEMENTS = 'UPDATE_SCORED_ACHIEVEMENTS';
export const UPDATE_SETTINGS_ACHIEVEMENTS = 'UPDATE_SETTINGS_ACHIEVEMENTS';
export const UPDATE_TIMED_ACHIEVEMENTS = 'UPDATE_TIMED_ACHIEVEMENTS';

export const RESET_CHANGED_ACHIEVEMENT = 'RESET_CHANGED_ACHIEVEMENT';
export const RESET_ACHIEVEMENTS = 'RESET_ACHIEVEMENTS';
