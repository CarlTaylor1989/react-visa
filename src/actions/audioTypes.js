export const SET_AUDIO = 'SET_AUDIO';
export const SET_PLAY_AUDIO = 'SET_PLAY_AUDIO';
export const PLAY_AUDIO = 'PLAY_AUDIO';
export const PAUSE_AUDIO = 'PAUSE_AUDIO';
export const STOP_AUDIO = 'STOP_AUDIO';
