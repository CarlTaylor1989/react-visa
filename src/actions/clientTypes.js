export const SET_CLIENT_PROPOSAL_DATA = 'SET_CLIENT_PROPOSAL_DATA';
export const SET_CLIENT_SHOWING = 'SET_CLIENT_SHOWING';
export const SET_ALL_CLIENTS = 'SET_ALL_CLIENTS';
export const SET_CLIENT_STATUS = 'SET_CLIENT_STATUS';
export const SET_PROPOSAL_AS_COMPLETE = 'SET_PROPOSAL_AS_COMPLETE';

export const INCREASE_REQUIREMENT_ATTEMPTS = 'INCREASE_REQUIREMENT_ATTEMPTS';
export const SET_REQUIREMENT_CORRECTNESS = 'SET_REQUIREMENT_CORRECTNESS';
export const RESET_REQUIREMENT_ATTEMPTS = 'RESET_REQUIREMENT_ATTEMPTS';
export const SET_REQUIREMENT_FEEDBACK = 'SET_REQUIREMENT_FEEDBACK';
export const DISPLAY_REQUIREMENT_FEEDBACK = 'DISPLAY_REQUIREMENT_FEEDBACK';
export const HIDE_REQUIREMENT_FEEDBACK = 'HIDE_REQUIREMENT_FEEDBACK';

export const SET_CLIENT_LOCKED = 'SET_CLIENT_LOCKED';
export const UNLOCK_CLIENT = 'UNLOCK_CLIENT';

export const DISPLAY_SOLUTION_FEEDBACK = 'DISPLAY_SOLUTION_FEEDBACK';
export const HIDE_SOLUTION_FEEDBACK = 'HIDE_SOLUTION_FEEDBACK';
export const SET_SOLUTION_FEEDBACK = 'SET_SOLUTION_FEEDBACK';
export const SET_SOLUTION_CORRECTNESS = 'SET_SOLUTION_CORRECTNESS';
export const INCREASE_SOLUTION_ATTEMPTS = 'INCREASE_SOLUTION_ATTEMPTS';
export const SET_SOLUTION_HAS_GIVEN_UP = 'SET_SOLUTION_HAS_GIVEN_UP';
export const SET_SOLUTION_SCORE = 'SET_SOLUTION_SCORE';
