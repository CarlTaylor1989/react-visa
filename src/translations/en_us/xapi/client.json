{
  "questionTitle": "%{name} client proposal step %{index}",
  "questionDescription": "%{name} client proposal with topic %{problem}, %{question} selection:",
  "question1": "statement",
  "question2": "product",

  "completedDescription": "A client proposal with topic %{problem}"
}
