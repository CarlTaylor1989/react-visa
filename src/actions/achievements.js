import * as types from './achievementTypes';

export function setCompletedAchievement(achievement) {
  return {
    type: types.SET_COMPLETED_ACHIEVEMENT,
    achievement
  };
}

export function restoreAchievements(completed, lastCompleted) {
  return {
    type: types.RESTORE_ACHIEVEMENTS,
    completed,
    lastCompleted
  };
}

export function updateProductAchievements() {
  return {
    type: types.UPDATE_PRODUCT_ACHIEVEMENTS
  };
}

export function updateProposalAchievements() {
  return {
    type: types.UPDATE_PROPOSAL_ACHIEVEMENTS
  };
}

export function updateScoredAchievements() {
  return {
    type: types.UPDATE_SCORED_ACHIEVEMENTS
  };
}

export function updateSettingsAchievements() {
  return {
    type: types.UPDATE_SETTINGS_ACHIEVEMENTS
  };
}

export function updateTimedAchievements() {
  return {
    type: types.UPDATE_TIMED_ACHIEVEMENTS
  };
}

export function resetChangedAchievement() {
  return {
    type: types.RESET_CHANGED_ACHIEVEMENT
  };
}

export function resetAchievements() {
  return {
    type: types.RESET_ACHIEVEMENTS
  };
}
