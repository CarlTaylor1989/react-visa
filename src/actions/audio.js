import * as types from './audioTypes';

export const setAudio = audioId => ({
  type: types.SET_AUDIO,
  audioId
});

export const setPlayAudio = audioId => ({
  type: types.SET_PLAY_AUDIO,
  audioId
});

export const playAudio = () => ({
  type: types.PLAY_AUDIO
});

export const pauseAudio = () => ({
  type: types.PAUSE_AUDIO
});

export const stopAudio = () => ({
  type: types.STOP_AUDIO
});
