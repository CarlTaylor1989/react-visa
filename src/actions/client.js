import * as types from './clientTypes';

export const setClientProposalData = clientProposalData => ({
  type: types.SET_CLIENT_PROPOSAL_DATA,
  clientProposalData
});

export const setClientShowing = clientShowing => ({
  type: types.SET_CLIENT_SHOWING,
  clientShowing
});

export const setAllClients = clients => ({
  type: types.SET_ALL_CLIENTS,
  clients
});

export const setClientStatus = (client, status) => ({
  type: types.SET_CLIENT_STATUS,
  client,
  status
});

export const increaseRequirementAttempts = client => ({
  type: types.INCREASE_REQUIREMENT_ATTEMPTS,
  client
});

export const setRequirementCorrectness = client => ({
  type: types.SET_REQUIREMENT_CORRECTNESS,
  client
});

export const resetRequirementAttempts = client => ({
  type: types.RESET_REQUIREMENT_ATTEMPTS,
  client
});

export const displayRequirementFeedback = () => ({
  type: types.DISPLAY_REQUIREMENT_FEEDBACK
});

export const hideRequirementFeedback = () => ({
  type: types.HIDE_REQUIREMENT_FEEDBACK
});

export const setRequirementFeedback = correct => ({
  type: types.SET_REQUIREMENT_FEEDBACK,
  correct
});

export const setLockedClient = (client, start, duration) => ({
  type: types.SET_CLIENT_LOCKED,
  client,
  start,
  duration
});

export const unlockClient = client => ({
  type: types.UNLOCK_CLIENT,
  client
});

export const setProposalAsComplete = client => ({
  type: types.SET_PROPOSAL_AS_COMPLETE,
  client
});

export const setSolutionFeedback = correct => ({
  type: types.SET_SOLUTION_FEEDBACK,
  correct
});

export const setSolutionCorrectness = client => ({
  type: types.SET_SOLUTION_CORRECTNESS,
  client
});

export const displaySolutionFeedback = () => ({
  type: types.DISPLAY_SOLUTION_FEEDBACK
});

export const hideSolutionFeedback = () => ({
  type: types.HIDE_SOLUTION_FEEDBACK
});

export const increaseSolutionAttempts = client => ({
  type: types.INCREASE_SOLUTION_ATTEMPTS,
  client
});

export const setSolutionScore = (client, score) => ({
  type: types.SET_SOLUTION_SCORE,
  client,
  score
});
