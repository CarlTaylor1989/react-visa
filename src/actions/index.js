import { push } from 'connected-react-router';
import * as AchievementActions from './achievements';
import * as AudioActions from './audio';
import * as ClientActions from './client';
import * as DiagnosticActions from './diagnostic';
import * as GenericActions from './generic';
import * as HintsActions from './hints';
import * as MapActions from './map';
import * as PromptActions from './prompts';
import * as ScoreActions from './score';
import * as ScormActions from './scorm';
import * as SettingsActions from './settings';
import * as StreakActions from './streak';
import * as TutorialActions from './tutorial';

const ActionCreators = Object.assign(
  {},
  AchievementActions,
  AudioActions,
  ClientActions,
  DiagnosticActions,
  GenericActions,
  HintsActions,
  MapActions,
  PromptActions,
  ScoreActions,
  ScormActions,
  SettingsActions,
  StreakActions,
  TutorialActions,
  { push }
);

export default ActionCreators;
