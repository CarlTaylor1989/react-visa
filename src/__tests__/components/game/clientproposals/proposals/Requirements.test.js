/* global toJson */
import React from 'react'; // eslint-disable-line no-unused-vars
import {
  Requirements,
  mapDispatchToProps
} from '../../../../../components/game/clientproposals/proposals/Requirements';
import * as actions from '../../../../../actions/client';
import * as types from '../../../../../actions/clientTypes';
import {
  checkStatementCorrectness,
  getNumberOfCorrectStatements,
  getStatementStatus,
  getRequirementStatements,
  getRequirementStatementIds
} from '../../../../../lib/clients';

jest.mock('../../../../../lib/clients');

let mockProps;
const statementId = 'stm1';
beforeEach(() => {
  mockProps = {
    answeredCorrectly: false,
    client: 'client1',
    setRequirementFeedback: jest.fn(),
    displayReqFeedback: true,
    closeResourcePopup: jest.fn()
  };

  getStatementStatus.mockImplementation(() => ({
    disabled: false,
    selected: false
  }));

  getRequirementStatementIds.mockImplementation(() => [statementId]);

  getRequirementStatements.mockImplementation(() => ([{
    id: statementId,
    text: 'irrelevant',
    type: 'opportunities'
  }]));
});

test('renders the Requirements component with the submit button', () => {
  const wrapper = shallow(<Requirements {...mockProps} />);
  expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
  expect(wrapper.instance().timeAvailable).toBeDefined();
});

test('renders the Requirements component without the submit button', () => {
  const wrapper = shallow(<Requirements {...mockProps} answeredCorrectly />);
  expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
});

describe('method componentDidUpdate()', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Requirements {...mockProps} />);
    wrapper.instance().setCheckboxesSubmit = jest.fn();
  });

  test('feedback panel is still open', () => {
    wrapper.setProps({ displayReqFeedback: true });
    expect(wrapper.instance().setCheckboxesSubmit).not.toHaveBeenCalled();
  });

  test('feedback panel closed and submit is enabled', () => {
    // setup
    checkStatementCorrectness.mockImplementation(() => true);
    wrapper.instance().state.checkboxes[statementId].selected = true;

    // function under test
    wrapper.setProps({ displayReqFeedback: false });

    // expectation
    expect(wrapper.instance().setCheckboxesSubmit).toHaveBeenCalledWith(true, expect.any(Object));
  });

  test('feedback panel closed and submit is disabled', () => {
    // setup
    checkStatementCorrectness.mockImplementation(() => false);
    wrapper.instance().state.checkboxes[statementId].selected = true;

    // function under test
    wrapper.setProps({ displayReqFeedback: false });

    // expectation
    expect(wrapper.instance().setCheckboxesSubmit).toHaveBeenCalledWith(false, expect.any(Object));
  });
});

describe('method onCheckboxChange()', () => {
  test('select statement', () => {
    // setup
    const wrapper = shallow(<Requirements {...mockProps} />);

    // function under test
    wrapper.instance().onCheckboxChange(statementId);

    // expectation
    expect(wrapper.instance().state.checkboxes[statementId]).toEqual(
      expect.objectContaining({
        selected: true
      })
    );
  });

  test('unselect statement', () => {
    // setup
    const checkboxStatementId = 'irrelevant';
    const wrapper = shallow(<Requirements {...mockProps} />);
    wrapper.instance().state.checkboxes[checkboxStatementId] = { selected: true };

    // function under test
    wrapper.instance().onCheckboxChange(checkboxStatementId);

    // expectation
    expect(wrapper.instance().state.checkboxes[checkboxStatementId]).toEqual(
      expect.objectContaining({
        selected: false
      })
    );
  });
});

describe('method onSubmit()', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Requirements {...mockProps} />);
    wrapper.instance().setCheckboxesSubmit = jest.fn();
    checkStatementCorrectness.mockImplementation(() => true);
  });

  test('incorrect statement and not selected', () => {
    // setup
    wrapper.instance().state.checkboxes[statementId].selected = false;
    getNumberOfCorrectStatements.mockImplementation(() => 2);

    // function under test
    wrapper.instance().onSubmit();

    // expectations
    expect(wrapper.instance().setCheckboxesSubmit).toHaveBeenCalledWith(false, expect.any(Object));
    expect(mockProps.setRequirementFeedback).toHaveBeenCalledWith(false);
  });

  test('incorrect statement and selected', () => {
    // setup
    wrapper.instance().state.checkboxes[statementId].selected = true;
    checkStatementCorrectness.mockImplementation(() => false);
    getNumberOfCorrectStatements.mockImplementation(() => 2);

    // function under test
    wrapper.instance().onSubmit();

    // expectations
    expect(wrapper.instance().setCheckboxesSubmit).toHaveBeenCalledWith(false, expect.any(Object));
    expect(mockProps.setRequirementFeedback).toHaveBeenCalledWith(false);
  });

  test('correct statement', () => {
    // setup
    wrapper.instance().state.checkboxes[statementId].selected = true;
    getNumberOfCorrectStatements.mockImplementation(() => 1);

    // function under test
    wrapper.instance().onSubmit();

    // expectations
    expect(wrapper.instance().setCheckboxesSubmit).toHaveBeenCalledWith(false, expect.any(Object));
    expect(mockProps.setRequirementFeedback).toHaveBeenCalledWith(true);
  });
});

test('method setCheckboxesSubmit()', () => {
  const wrapper = shallow(<Requirements {...mockProps} />);
  wrapper.instance().setCheckboxesSubmit(false, wrapper.instance().state.checkboxes);
  expect(wrapper.instance().state.canSubmit).toBeFalsy();
  expect(wrapper.instance().state.checkboxes).toEqual(expect.any(Object));
});

test('should mapDispatchToProps', () => {
  const correct = false;
  const dispatch = jest.fn();
  jest.spyOn(actions, 'setRequirementFeedback').mockReturnValue('setRequirementFeedback()');
  const props = mapDispatchToProps(dispatch);
  props.setRequirementFeedback(correct);
  expect(dispatch).toHaveBeenCalledWith({
    type: types.SET_REQUIREMENT_FEEDBACK,
    correct
  });
});
