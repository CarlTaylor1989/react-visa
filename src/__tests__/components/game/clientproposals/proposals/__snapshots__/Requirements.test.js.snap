// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Requirements component with the submit button 1`] = `
<div
  className="requirements"
>
  <div
    className="group"
  >
    <StatementsInstruction
      client="client1"
      completed={false}
    />
    <div
      className="scrollingArea"
    >
      <Scrollbar
        contentProps={Object {}}
        fallbackScrollbarWidth={20}
        maximalThumbYSize={83}
        minimalThumbSize={30}
        minimalThumbYSize={83}
        momentum={true}
        noDefaultStyles={true}
        scrollDetectionThreshold={100}
        scrollerProps={Object {}}
        thumbXProps={Object {}}
        thumbYProps={Object {}}
        trackClickBehavior="jump"
        trackXProps={Object {}}
        trackYProps={Object {}}
        wrapperProps={Object {}}
      >
        <div
          className="statementsWrapper"
        >
          <StatementsGroup
            client="client1"
            group="opportunities"
            key="opportunities"
          />
          <CheckboxLabel
            disabled={false}
            elementId="stm1"
            isSelected={false}
            key="opportunitiesstm10"
            label="irrelevant"
            onCheckboxChange={[Function]}
          />
        </div>
      </Scrollbar>
    </div>
  </div>
  <button
    className="submitBtn"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    <Translate
      tag="span"
      value="clientprops.submitBtn"
    />
  </button>
</div>
`;

exports[`renders the Requirements component without the submit button 1`] = `
<div
  className="requirements completed"
>
  <div
    className="group"
  >
    <StatementsInstruction
      client="client1"
      completed={true}
    />
    <div
      className="scrollingArea"
    >
      <Scrollbar
        contentProps={Object {}}
        fallbackScrollbarWidth={20}
        maximalThumbYSize={83}
        minimalThumbSize={30}
        minimalThumbYSize={83}
        momentum={true}
        noDefaultStyles={true}
        scrollDetectionThreshold={100}
        scrollerProps={Object {}}
        thumbXProps={Object {}}
        thumbYProps={Object {}}
        trackClickBehavior="jump"
        trackXProps={Object {}}
        trackYProps={Object {}}
        wrapperProps={Object {}}
      >
        <div
          className="statementsWrapper"
        >
          <StatementsGroup
            client="client1"
            group="opportunities"
            key="opportunities"
          />
          <CheckboxLabel
            disabled={false}
            elementId="stm1"
            isSelected={false}
            key="opportunitiesstm10"
            label="irrelevant"
            onCheckboxChange={[Function]}
          />
        </div>
      </Scrollbar>
    </div>
  </div>
</div>
`;
